hard code buider image for buildpack and change default builder image

xiujuan95 opened the pull request:
Fix issue: https://github.com/redhat-developer/build/issues/140
- Hard code builder image for buildpack build strategy to `heroku/buildpacks:18`;
- Change default builder image according to the provided builder image in build spec;
This is because we don't have a common builder image for all build strategies. So we shouldn't hard code the default image. 

Now, we hard code builder image for all build strategies. So if now, we don't provide a builder image in build spec, then they will use the builder image within strategy definition. 

But for `source-to-image` strategy, It's different. For it, we must give a builder image in build spec. Because for it, the builder image in here :https://github.com/redhat-developer/build/blob/master/samples/buildstrategy/source-to-image/buildstrategy_source-to-image_cr.yaml#L14
It's a based image, for different source code, it should have different builder images. Such as `nodejs` and `java` should have different builder image.
Now, if you don't give a builder image in build spec for `source-to-image`, then it will fail. This is expected.

- Keep `spec.builder.image` is optional.

Within buildpack build spec, there is no builder image is defined:
```
apiVersion: build.dev/v1alpha1
kind: Build
metadata:
  name: buildpack-nodejs-build
spec:
  source:
    url: https://github.com/sclorg/nodejs-ex
  strategy:
    name: buildpacks-v3
    kind: ClusterBuildStrategy
  output:
    image: us.icr.io/zoe_namespace/buildpack-example
```
And then it will use the image defined within strategy:
```
spec:
  inputs:
    resources:
    - name: source
      resourceSpec:
        params:
        - name: url
          value: https://github.com/sclorg/nodejs-ex
        - name: revision
          value: master
        type: git
  outputs:
    resources:
    - name: image
      resourceSpec:
        params:
        - name: url
          value: us.icr.io/zoe_namespace/buildpack-example
        type: image
  serviceAccountName: pipeline
  taskSpec:
    inputs:
      params:
      - default: Dockerfile
        description: Path to the Dockerfile
        name: DOCKERFILE
        type: string
      - default: .
        description: The root of the code
        name: PATH_CONTEXT
        type: string
      resources:
      - name: source
        type: git
    outputs:
      resources:
      - name: image
        type: image
    steps:
    - args:
      - -c
      - chown -R "1000:1000" "/workspace/source"
      command:
      - /bin/bash
      image: heroku/buildpacks:18
      name: step-prepare
```
The status of `taskrun`:
```
kubectl get tr
NAME                              SUCCEEDED   REASON      STARTTIME   COMPLETIONTIME
buildpack-nodejs-buildrun-hzdtw   True        Succeeded   14m         13m
```
Run e2e test is passed:
```
INFO[0000] Testing operator locally.                    
time="2020-04-27T04:09:01-04:00" level=info msg="Started local operator"
=== RUN   TestBuild
=== RUN   TestBuild/build-group
=== RUN   TestBuild/build-group/Build_e2e_tests
=== PAUSE TestBuild/build-group/Build_e2e_tests
=== CONT  TestBuild/build-group/Build_e2e_tests
--- PASS: TestBuild (535.67s)
    --- PASS: TestBuild/build-group (0.00s)
        --- PASS: TestBuild/build-group/Build_e2e_tests (523.38s)
            main_test.go:69: Initialized cluster resources
            wait_util.go:48: Operator is running locally; skip waitForDeployment
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/kaniko) created
            client.go:62: resource type Build with namespace/name (default/example-build-kaniko) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-kaniko) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/kaniko) created
            client.go:62: resource type Build with namespace/name (default/kaniko-custom-context-dockerfile) created
            client.go:62: resource type BuildRun with namespace/name (default/kaniko-custom-context-dockerfile) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/source-to-image) created
            client.go:62: resource type Build with namespace/name (default/example-build-s2i) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-s2i) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/buildah) created
            client.go:62: resource type Build with namespace/name (default/example-build-buildah) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-buildah) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/buildah) created
            client.go:62: resource type Build with namespace/name (default/buildah-custom-context-dockerfile) created
            client.go:62: resource type BuildRun with namespace/name (default/buildah-custom-context-dockerfile) created
            client.go:62: resource type ClusterBuildStrategy with namespace/name (/buildpacks-v3) created
            client.go:62: resource type Build with namespace/name (default/example-build-buildpacks-v3) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-buildpacks-v3) created
            client.go:62: resource type BuildStrategy with namespace/name (default/buildpacks-v3) created
            client.go:62: resource type Build with namespace/name (default/example-build-buildpacks-v3-namespaced) created
            client.go:62: resource type BuildRun with namespace/name (default/example-build-buildpacks-v3-namespaced) created
            client.go:82: resource type BuildRun with namespace/name (default/example-build-buildpacks-v3-namespaced) successfully deleted
            client.go:82: resource type Build with namespace/name (default/example-build-buildpacks-v3-namespaced) successfully deleted
            client.go:82: resource type BuildStrategy with namespace/name (default/buildpacks-v3) successfully deleted
            client.go:82: resource type BuildRun with namespace/name (default/example-build-buildpacks-v3) successfully deleted
            client.go:82: resource type Build with namespace/name (default/example-build-buildpacks-v3) successfully deleted
            client.go:82: resource type ClusterBuildStrategy with namespace/name (/buildpacks-v3) successfully deleted
            client.go:82: resource type BuildRun with namespace/name (default/buildah-custom-context-dockerfile) successfully deleted
            client.go:82: resource type Build with namespace/name (default/buildah-custom-context-dockerfile) successfully deleted
            client.go:82: resource type ClusterBuildStrategy with namespace/name (/buildah) successfully deleted
            client.go:82: resource type BuildRun with namespace/name (default/example-build-buildah) successfully deleted
            client.go:82: resource type Build with namespace/name (default/example-build-buildah) successfully deleted
            client.go:82: resource type ClusterBuildStrategy with namespace/name (/buildah) successfully deleted
            client.go:82: resource type BuildRun with namespace/name (default/example-build-s2i) successfully deleted
            client.go:82: resource type Build with namespace/name (default/example-build-s2i) successfully deleted
            client.go:82: resource type ClusterBuildStrategy with namespace/name (/source-to-image) successfully deleted
            client.go:82: resource type BuildRun with namespace/name (default/kaniko-custom-context-dockerfile) successfully deleted
            client.go:82: resource type Build with namespace/name (default/kaniko-custom-context-dockerfile) successfully deleted
            client.go:82: resource type ClusterBuildStrategy with namespace/name (/kaniko) successfully deleted
            client.go:82: resource type BuildRun with namespace/name (default/example-build-kaniko) successfully deleted
            client.go:82: resource type Build with namespace/name (default/example-build-kaniko) successfully deleted
            client.go:82: resource type ClusterBuildStrategy with namespace/name (/kaniko) successfully deleted
PASS
......
ok      github.com/redhat-developer/build/test/e2e      537.146s
INFO[0540] Local operator test successfully completed.
```
And nit test is also passed:
```
oot@jordan-0:~/gopath/src/build/pkg/controller/buildrun# go test
PASS
ok      github.com/redhat-developer/build/pkg/controller/buildrun       0.275s
```

zhangtbj commented:
Hi @xiujuan95 ,

Let us confirm with the team if we all agree this solution. 

If yes, you also need to remove the builder.image from two buildpacks builds:
https://github.com/redhat-developer/build/blob/master/samples/build/build_buildpacks-v3_cr.yaml

And the builder image in the readme.md:
https://github.com/redhat-developer/build/blob/master/README.md

otaviof commented:
/ok-to-test

xiujuan95 commented:
@zhangtbj Thanks for your review! I have removed `spec.builder.image` in buildpack build sample and README.md. 

And also removed builder image in test/data buildpack sample: https://github.com/redhat-developer/build/blob/master/test/data/build_buildpacks-v3_cr_private_github.yaml
Thanks @qu1queee for your kind reminder! 

Besides, I have pulled the latest code and fix conflicts and travis also has passed:https://travis-ci.com/github/redhat-developer/build/builds/162857072

sbose78 commented:
Direction looks good!

https://github.com/redhat-developer/build/issues/140#issuecomment-621906997

Are we failing the build if some buildstrategy needs the builder image specified but wasn't provided in the Build CR?

sbose78 commented on pkg/controller/buildrun/generate_taskrun.go:
```diff
@@ -99,6 +91,18 @@ func generateTaskSpec(build *buildv1alpha1.Build, buildRun *buildv1alpha1.BuildR
 	}
 
+	if build.Spec.BuilderImage != nil {
+		InputBuilderImage := v1alpha2.ParamSpec {
```
This is worrying me a bit, because we are ending up leaking abstraction by mapping something in the `build` CR to a tekton object directly.

zhangtbj replied:
Hi @sbose78 ,

Do you have any comment about this review?

Do you mean it is not good to map the `InputBuilderImage` as tekton object directly?

Thanks!

sbose78 replied:
I think this is fine!

sbose78 replied:
Ideally, the builder image should go into the Tekton object only if it was referred to in the BuildStrategy, right ?

xiujuan95 replied:
@sbose78 Yes, I think so!

sbose78 replied:
I don't think we had that 'check' in the first place, we can skip it for now. I'll let you two weigh in if that needs to be addressed later on :)

xiujuan95 replied:
I think we need to check it in the first place. Because for source-to-image buildStrategy, we must provide a builder image definition in build spec. Because for different source code languages, we need different builder images.

otaviof commented:
I think in general, hard-coding the builder image in the operator would not help. Since:
- would require a operator release in order to update those CNBs
- would create hard commitment with a single CNB image, something that buildpacks was design to accept any image that follows the [lifecycle specification](https://buildpacks.io/docs/concepts/components/lifecycle/)
- would render `spec.builder.image` useless for buildpacks strategy, without a clear alert for the user

With that said, I think the route we should follow is to validate the image informed by the user. Check if it complies with buildpacks lifecycle specification supported by the build-operator, and if the validation fails, we can create a clear error message for the user. As buildpacks would evolve, we can follow up on build-operator supporting newer lifecycle specifications, while keeping the build concise.

What's your take on it?

zhangtbj commented:
Hi @otaviof ,

I think this PR addressed two requirements:

- A bug if we didn't config the builder.image in build yaml: https://github.com/redhat-developer/build/issues/146
- A requirement for the customer: https://github.com/redhat-developer/build/issues/148

I think just few end-users know what builder image or builder image version they want to use, I think most of the end-user just know I am using Dockerfile or buildpacks, and I want to build my image, so they don't care about which builder they use or we provide.

So I also suggest in each release we announce for example, we already verify cloudfoundry builder image version x.x.x works fine in our latest vx.y.z release. 

I agree all buildpacks image should work correctly and same which follow buildpacks lifecycle standard. But maybe it is not the real case, we are verifying the paketo buildpacks, maybe the exporter cannot work well as before, maybe because the special setting or maybe because the lifecycle version is different, like: (pack inspect-builder gcr.io/paketo-buildpacks/builder vs pack inspect-builder heroku/buildpacks:18: 
`v0.7.3` vs `v0.6.1`)

We cannot make sure all versions, all builder images, all parameters can work fine within one build strategy (And we cannot verify them all).

Or at least we should provide the default builder image if end-user doesn't set it, but we have to hard code the builder image version in the go code, it is worse.

Hope it is helpful :)

zhangtbj commented:
Hi @xiujuan95 ,

Pls also provide us the result (the only s2i buildstrategy case) if we don't set the builder.image in the build definition.

We also need to make sure we can handle that well and the failure result is good for us. Thanks!

xiujuan95 commented:
@sbose78 Thanks for your review! For this question, answer is yes!

> Direction looks good!
> 
> [#140 (comment)](https://github.com/redhat-developer/build/issues/140#issuecomment-621906997)
> 
> Are we failing the build if some buildstrategy needs the builder image specified but wasn't provided in the Build CR?

For example, `source-to-image` buildStrategy, we must provide a builder image, otherwise, it will failed because below error:
```
admission webhook "validation.webhook.pipeline.tekton.dev" denied the request: validation failed: non-existent variable in "$(inputs.params.BUILDER_IMAGE)" for step command[3]: taskspec.steps.command[3]
```

@otaviof also thanks for your comments. I agree with @zhangtbj 's comment. I also think end-user doesn't care about what's the builder image and they just want to use this buildStrategy to build themselves' image. Most of the time, just developers care about builder image. 

So I think we need to hard code the builder image first. Of course, for `source-to-image`, although we hard code it in buildStrategy, we still need to provide another builder image. Because for different code language, it has different builder image. Under this situation, if user doesn't give build image parameter, then buildrun will fail. Below is an example:
```
kubectl get buildrun
NAME                           SUCCEEDED   REASON                                                                                                                                                                                                       STARTTIME   COMPLETIONTIME
s2i-nodejs-buildrun            True        Succeeded                                                                                                                                                                                                    3m31s       108s
s2i-nodejs-buildrun-no-image   False       admission webhook "validation.webhook.pipeline.tekton.dev" denied the request: validation failed: non-existent variable in "$(inputs.params.BUILDER_IMAGE)" for step command[3]: taskspec.steps.command[3]   5m10s
```
Hope above answers can help you! @sbose78 @otaviof @zhangtbj

sbose78 commented:
@otaviof Are you good with this?

qu1queee commented:
@sbose78 @otaviof any reasons not to merge this? this PR is quite old.

sbose78 commented:
Let's take it in

/lgtm
/approve
